src/mdns/mdns_avahi.cpp

```diff
@@ -1050,6 +1054,7 @@ void PublisherAvahi::ServiceSubscription::HandleBrowseResult(AvahiServiceBrowser
         break;
     case AVAHI_BROWSER_REMOVE:
         mPublisherAvahi->OnServiceRemoved(static_cast<uint32_t>(aInterfaceIndex), aType, aName);
+        RemoveServiceResolver(aName);
```

Comment:
Is it able to handle this scenario?

```
1. service A is published.
2. service A is unpublished.
3. service A is published again.
4. service A is updated.
```

Reply:
The scenario works, tested with 2 devices, device 1 with trel disabled, device 2 with trel enabled:
1. enable trel on device 1, service A is published and resolved by device 2
2. disable trel on device 1, service A is removed and resolved by device 2
3. enable trel on device 1, service A is added and resolved by device 2
4. reboot device 1, service A is updated and resolved by device 2

logs on device B:
```
1)
03-27 02:21:19.969  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.977  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.985  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.994  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:21:19.995  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1
03-27 02:21:20.000  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1

2)
03-27 02:22:16.881  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.897  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.897  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.907  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.907  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.907  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c
03-27 02:22:16.908  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 02:22:16.908  2013  2013 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c

3)
03-27 02:22:44.642  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.651  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.658  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.666  2013  2013 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 02:22:44.666  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1
03-27 02:22:44.671  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 3EAB46586584.local. addresses 1

4)
03-27 02:24:59.952  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host CA1750BED820.local. addresses 1
03-27 02:24:59.959  2013  2013 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host CA1750BED820.local. addresses 1
```

Reply:
Offline discussed with @superwhd , `avahi_service_browser_new` uses `AVAHI_PROTO_UNSPEC ` while `avahi_service_resolver_new` uses `AVAHI_PROTO_INET6 `, there might be cases when browse callback was called for v4 event while resolver is created with v6 only.

Updated `avahi_service_resolver_new` to use `AVAHI_PROTO_UNSPEC ` to align with browser, and the logs for above scenario becomes:

```
1)
03-27 10:01:45.311  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:01:45.319  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:01:45.326  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 2E05397250BE-2.local. addresses 1

2)
03-27 10:02:13.548  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 10:02:13.556  1986  1986 D otbr-agent: [DEBG]-MDNS----: Removed 2 service resolver for instance 12e24f966ff2295c
03-27 10:02:13.556  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: remove 12e24f966ff2295c host  addresses 0
03-27 10:02:13.557  1986  1986 D otbr-agent: [DEBG]-MDNS----: Removed 0 service resolver for instance 12e24f966ff2295c

3)
03-27 10:02:19.909  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:02:19.917  1986  1986 D otbr-agent: [DEBG]-MDNS----: Added service resolver for instance 12e24f966ff2295c
03-27 10:02:19.922  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 2E05397250BE-2.local. addresses 1

4)
03-27 10:03:39.284  1986  1986 I otbr-agent: [INFO]-MDNS----: Service _trel._udp is resolved successfully: add 12e24f966ff2295c host 9ED50C0C67C0.local. addresses 1
```

Reply:
> Offline discussed with @superwhd , avahi_service_browser_new uses AVAHI_PROTO_UNSPEC  while avahi_service_resolver_new uses AVAHI_PROTO_INET6 , there might be cases when browse callback was called for v4 event while resolver is created with v6 only.
>
>Updated avahi_service_resolver_new to use AVAHI_PROTO_UNSPEC  to align with browser, and the logs for above scenario becomes:

+1, it's recommended to browse/resolve with both IPv4 and IPv6 per the Discovery Proxy spec:
```
At present, it is RECOMMENDED that a Discovery Proxy be configured with a single domain name for both
the IPv4 and IPv6 ".local." zones on the local link, and when a unicast query is received, it should issue
Multicast DNS queries using both IPv4 and IPv6 on the local link and then combine the results.
```